scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/context/TaskContextVariableExtractor.java

```diff
@@ -0,0 +1,144 @@
+ * USA
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ *
+ *  Initial developer(s):               The ProActive Team
+ *                        http://proactive.inria.fr/team_members.htm
+ *  Contributor(s):
+ *
+ *  * $$ACTIVEEON_INITIAL_DEV$$
+ */
+package org.ow2.proactive.scheduler.task.context;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.ow2.proactive.scheduler.common.task.ForkEnvironment;
+import org.ow2.proactive.scheduler.common.task.TaskResult;
+import org.ow2.proactive.scheduler.common.task.util.SerializationUtil;
+import org.ow2.proactive.scheduler.common.util.VariableSubstitutor;
+import org.ow2.proactive.scheduler.task.SchedulerVars;
+import org.ow2.proactive.scheduler.task.TaskLauncherInitializer;
+import org.ow2.proactive.scheduler.task.executors.forked.env.ForkedTaskVariablesManager;
+import org.jetbrains.annotations.NotNull;
+
+public class TaskContextVariableExtractor implements Serializable {
+    private final ForkedTaskVariablesManager forkedTaskVariablesManager = new ForkedTaskVariablesManager();
+
+    public Map<String, String> extractVariablesThirdPartyCredentialsAndSystemEnvironmentVariables(TaskContext taskContext) throws Exception {
+        ForkEnvironment forkEnvironment = taskContext.getInitializer().getForkEnvironment();
+        Map<String, Serializable> variables = this.extractTaskVariables(taskContext);
+        Map<String, String> thirdPartyCredentials = forkedTaskVariablesManager.extractThirdPartyCredentials(taskContext);
+        HashMap<String, Serializable> systemEnvironmentVariables = new HashMap<String, Serializable>(
+                System.getenv());
+        systemEnvironmentVariables.putAll(variables);
+        systemEnvironmentVariables.putAll(thirdPartyCredentials);
+
+        return VariableSubstitutor.filterAndUpdate(forkEnvironment.getSystemEnvironment(),
+                systemEnvironmentVariables);
+    }
+
+    public Map<String, Serializable> extractTaskVariables(TaskContext taskContext) throws Exception {
+        return extractTaskVariables(taskContext, (TaskResult) null);
+    }
+
+    private Map<String, Serializable> extractTaskVariables(TaskContext container, TaskResult taskResult,
+            String nodesFile)
+            throws Exception {
+        Map<String, Serializable> variables = extractTaskVariables(container, taskResult);
+
+        variables.put(SchedulerVars.PA_NODESNUMBER.toString(), container.getOtherNodesURLs().size() + 1);
+        variables.put(SchedulerVars.PA_NODESFILE.toString(), nodesFile);
+
+        variables.put(SchedulerVars.PA_TASK_PROGRESS_FILE.toString(), container.getProgressFilePath());
+
+        return variables;
+    }
+
+    public Map<String, Serializable> extractTaskVariables(TaskContext taskContext,
+            TaskResult taskResult) throws Exception {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)